### PR TITLE
Fix cursor automatically moving to next position when delete and put are in the same TXN

### PR DESCRIPTION
**Pull request by ahrtr**
Fix https://github.com/boltdb/bolt/issues/357

cc @thatguystone @flowchartsman @benbjohnson

It's a bit funny that a bug isn't fixed until 8 years later. 

Just copied the comment to explain the root cause,

```
	// If the deleted item is the first item in the node, we should keep
	// the cursor's position (decrease by 1). Otherwise, the cursor will
	// automatically move to the next item, and when clients call `Next`
	// method afterward, it will skip one item in the node.
	//
	// Note if there isn't any write on current page in current transaction,
	// then the cursor only iterates items against the page instead of the
	// in-memory node. Since the page is immutable, so we don't need to change
	// pos.index in such case.
```

### CODE REVIEW

**Comment by ahrtr** [comment]
cc @ptabor @fuweid @tjungblu as well

File: cursor_test.go [hunk 1]
```diff
@@ -742,6 +745,95 @@ func TestCursor_QuickCheck_BucketsOnly_Reverse(t *testing.T) {
...
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			db := btesting.MustCreateDB(t)
+
+			err := db.Update(func(tx *bolt.Tx) error {
+				b, _ := tx.CreateBucket([]byte("widgets"))
+				for i := 0; i < count; i++ {
+					k := make([]byte, 8)
+					binary.BigEndian.PutUint64(k, uint64(i))
+					if perr := b.Put(k, make([]byte, 100)); perr != nil {
```
Comment (fuweid) [inline cursor_test.go:793]: suggest adding matric with different value size to construct different stack length for test.

**Review by fuweid — COMMENTED** [review]
The change looks good.

File: cursor_test.go (hunk 1 above)
Comment (fuweid) [reply cursor_test.go:793]: Different value size is for separate Txn after rebalancing.

File: cursor.go [hunk 2]
```diff
@@ -151,6 +151,20 @@ func (c *Cursor) Delete() error {
...
+	// the cursor's position (decrease by 1). Otherwise, the cursor will
+	// automatically move to the next item, and when clients call `Next`
+	// method afterward, it will skip one item in the node.
+	//
+	// Note if there isn't any write on current page in current transaction,
+	// then the cursor only iterates items against the page instead of the
+	// in-memory node. Since the page is immutable, so we don't need to change
+	// pos.index in such case.
+	if pos != nil && pos.index == 0 && pos.node != nil {
+		pos.index--
```
Comment (tjungblu) [inline cursor.go:165]: nit: that's personal preferences but I would just set this to -1 directly. Takes a little mental math off my head when reading.

File: cursor_test.go [hunk 3]
```diff
@@ -742,6 +745,95 @@ func TestCursor_QuickCheck_BucketsOnly_Reverse(t *testing.T) {
...
+							return err
+						}
+					}
+
+					return nil
+				})
+				require.NoError(t, err)
+			}
+
+			err = db.View(func(tx *bolt.Tx) error {
```
Comment (tjungblu) [inline cursor_test.go:827]: is it worth iterating the whole bucket with a cursor again and check all keys for correctness? relying on the stats might also prove to be wrong :)

File: cursor_test.go (hunk 1 above)
Comment (fuweid) [reply cursor_test.go:793]: IMO, yes. just want to verify it when the p.node is nil

File: cursor.go (hunk 2 above)
Comment (ahrtr) [reply cursor.go]: updated.

File: cursor_test.go (hunk 3 above)
Comment (ahrtr) [reply cursor_test.go:827]: The purpose is just to get the total number of key/value pair in the bucket. The `b.Stats().KeyN` is simple enough, and have better readability, although it also generates some other stats data which isn't needed in this test. Note that it's a readonly TXN, we can trust on the stats.

**Comment by ahrtr** [comment]
## Actually the fix isn't enough.  See the following two scenarios

Assuming we already inserted the following key/value pairs in the same TXN.
```
		for i := 0; i < count; i++ {
			k := make([]byte, 8)
			binary.BigEndian.PutUint64(k, uint64(i))
			b.Put(k, make([]byte, 100))
		}
```

- If clients starts deleting from an intermediate item instead of the first item all the way to the end. The number of remaining key/value pair is still incorrect. 

The example below starts deleting from the third item (with index 2), so the number of remaining key/value pair should be 2, but actually it isn't. **The reason is it skips one item each time after deleting one item.**

```
		c := b.Cursor()
		c.First()
		c.Next()
		key, _ := c.Next()
		for key != nil {
			if err := c.Delete(); err != nil {
				return err
			}
			key, _ = c.Next()
		}
```

- Assuming you want to remove consecutive N key/value pair starting from a certain key. It will not remove the expected key/value pairs. 

The example below removes consecutive 3 key/value pair starting from key `2`. We expect it  to remove key `2`, `3`, and `4`, but actually it removes `2`, `4` and `6`.  It's the same reason: `it skips one item each time after deleting one item`.
```
		c := b.Cursor()

		target := make([]byte, 8)
		binary.BigEndian.PutUint64(target, uint64(2))
		cnt := 3

		found := false
		for key, _ := c.First(); key != nil; key, _ = c.Next() {
			if bytes.Equal(key, target) {
				found = true
			}
			if found {
				if cnt > 0 {
					if err := c.Delete(); err != nil {
						return err
					}
					cnt--
				} else {
					break
				}
			}
		}
```

## Root cause and solution
Just as I mentioned above, the root cause is that the cursor is based on a mutable in-memory node. Each time when removing an item, the cursor automatically moves to next item.

The solution should be maintain a immutable node copy for stable cursor. But it may have some impact on the performance.

File: cursor_test.go (hunk 1 above)
Comment (ahrtr) [reply cursor_test.go:793]: Agree that we should add more test, but actually it has bigger concern https://github.com/etcd-io/bbolt/pull/611#issuecomment-1814879792

**Comment by fuweid** [comment]
Good catch! I did pulled your changes in my local and change the initPos with (c.First(), c.Next()) to test. But I didn't reproduce the issue 🥶

**Comment by ahrtr** [comment]
There is no cheap fix. Usually users won't delete key/value pairs either during iteration. So decided to just update doc to clarify the behavior.  https://github.com/etcd-io/bbolt/pull/614

**Comment by missinglink** [comment]
Agh funny, I just opened https://github.com/etcd-io/bbolt/pull/629, came to the same conclusions.

**Comment by ahrtr** [comment]
> Agh funny, I just opened #629, came to the same conclusions. Seems it's not as simple as I would have hoped 😢

Right, it isn't that simple. It doesn't deserve too much effort to fix it for now, unless lots of users drive it. Also it may either complicate the user interface or decrease the performance if we follow https://github.com/etcd-io/bbolt/pull/611#issuecomment-1817064812. 
- If we clone the `node.inodes` each time when constructing the `elemRef`, then it may cause reduce of performance and more usage of memory.
- If we provide an option to let users to decide whether to clone `node.inodes` each time when constructing the `elemRef`, it will complicate the user interface and also hard to understand.


Let's update the doc (see https://github.com/etcd-io/bbolt/pull/614) for now.

**Comment by ahrtr** [comment]
No plan to continue to work on this for now.

Note: the doc has already been updated in https://github.com/etcd-io/bbolt/pull/614

**Pull request closed without merging**